stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,168 @@
+        kernel_regularizer (str or func, optional): The regulariser to use for the weights of each layer.
+        kernel_constraint (str or func, optional): The constraint to use for the weights of each layer.
+        bias_initializer (str or func, optional): The initialiser to use for the bias of each layer.
+        bias_regularizer (str or func, optional): The regulariser to use for the bias of each layer.
+        bias_constraint (str or func, optional): The constraint to use for the bias of each layer.
+    """
+
+    def __init__(
+        self,
+        layer_sizes,
+        generator,
+        bias=True,
+        dropout=0.0,
+        activations=None,
+        kernel_initializer=None,
+        kernel_regularizer=None,
+        kernel_constraint=None,
+        bias_initializer=None,
+        bias_regularizer=None,
+        bias_constraint=None,
+    ):
+
+        super().__init__(
+            layer_sizes,
+            generator,
+            bias=bias,
+            dropout=dropout,
+            activations=activations,
+            kernel_initializer=kernel_initializer,
+            kernel_regularizer=kernel_regularizer,
+            kernel_constraint=kernel_constraint,
+            bias_initializer=bias_initializer,
+            bias_regularizer=bias_regularizer,
+            bias_constraint=bias_constraint,
+        )
+
+    def unsupervised_node_model(self):
+        """
+        A function to create the the inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+        Returns:
+            input and output layers for use with a keras model
+        """
+        # Inputs for features
+        x_t = Input(batch_shape=(1, self.n_nodes, self.n_features))
+        # Inputs for shuffled features
+        x_corr = Input(batch_shape=(1, self.n_nodes, self.n_features))
+
+        out_indices_t = Input(batch_shape=(1, None), dtype="int32")
+
+        # Create inputs for sparse or dense matrices
+        if self.use_sparse:
+            # Placeholders for the sparse adjacency matrix
+            A_indices_t = Input(batch_shape=(1, None, 2), dtype="int64")
+            A_values_t = Input(batch_shape=(1, None))
+            A_placeholders = [A_indices_t, A_values_t]
+
+        else:
+            # Placeholders for the dense adjacency matrix
+            A_m = Input(batch_shape=(1, self.n_nodes, self.n_nodes))
+            A_placeholders = [A_m]
+
+        x_inp = [x_corr, x_t, out_indices_t] + A_placeholders
+
+        node_feats = self([x_t, out_indices_t] + A_placeholders)
+        node_feats = Lambda(lambda x: K.squeeze(x, axis=0,), name="NODE_FEATURES",)(
+            node_feats
+        )
+
+        node_feats_corrupted = self([x_corr, out_indices_t] + A_placeholders)
+        node_feats_corrupted = Lambda(lambda x: K.squeeze(x, axis=0,),)(
+            node_feats_corrupted
+        )
+
+        summary = Lambda(lambda x: tf.math.sigmoid(tf.math.reduce_mean(x, axis=0)))(
```

Comment:
Is it worth having a `ReadOut` later similar to `Discriminator`?

Reply:
I feel like its simple enough to keep as a lambda, but I might create a `ReadOut` when I make the GraphSAGE infomax PR

Reply:
You might be able to use the `tf.keras.layers.GlobalAveragePooling1D` layer here.

Reply:
I'm hesitant to add this in here, `tf.keras.layers.GlobalAveragePooling1D` requires tensors with 3 dimension so won't work with mini-batch methods. It would also require keeping the batch dimension in . Squeezing out the batch dimension is convenient for calculating the scores and the embeddings. Also, squeezing out the batch dimension gives the node feature tensors the same number of dimension as the mini-batch methods which means `Discriminator` and `embedding_model` will work as is.

Reply:
I'm not sure if I understood correctly so let's work out the tensor dimensions.

The input `node_features` is (1, N, F) where N is the number of nodes in the graph and F is the dimensionality of the features. The first dimension, 1, is the batch size. In your case it is always 1 so you remove it to get an (N,F) tensor and then average across the rows to get an (F, ) tensor that is `summary`.  Your corrupted `node_features_corrupted` also has the first dimension removed so you have an (N, F) tensor. Your discriminator assumes that you are given an (N, F) and an (F, ) tensors which I think is a strong assumption if you want to support batch sizes greater than 1.

The `GlobalAveragePooling1D` will take as input your (1, N, F) tensor and the return (1, F). At this point, you can remove the first dimension and you have an (F, ) tensor for `summary` and you can continue as before. However, if you implement your `Discriminator` to work with inputs (batch size, N, F) and (batch size, F) then it will work for all batch sizes and you won't have to squeeze and then add the batch size. You should be able (I think) to re-write your `Discriminator` layer to use Keras `batch_dot` operator (https://www.tensorflow.org/api_docs/python/tf/keras/backend/batch_dot) and then you can pass tensors that don't have the batch dimension removed.

Reply:
> Why is squeezing out the batch dimension convenient for those things?

For most downstream tasks its useful to have embeddings in the shape `(number_of_nodes, dimension)` without the batch dimension of 1 e.g. plotting, clustering, training a neural network. So I feel like its more convenient for the user if we squeeze out the batch dim for them.

It also means that the embedding model will behave the same for graphsage type algorithms. In this case the embedding model will return tensors with shape `(head_nodes, dimension)`, so the user won't have to change any downstream code i.e. removing/adding batch dim squeezing code.

Re. it being convenient for the `Discriminator` @PantelisElinas makes a good point that it could re-written to handle both `(1, nodes, feats)` and `(head_nodes, feats)` input tensors quite easily. 

I think the `GlobalAveragePooling1D` won't work for `GraphSAGE` but it makes sense to worry about `GraphSAGE` implementation when I'm actually implementing it! I'll add `GlobalAveragePooling1D`, and maybe still squeeze out the batch dim for the embedding model. Thoughts?